src/azure-cli/azure/cli/command_modules/appconfig/_kv_helpers.py

```diff
@@ -120,43 +120,59 @@ def validate_import_feature(feature):
 
 # File <-> List of KeyValue object
 
-
-def __read_kv_from_file(file_path,
-                        format_,
-                        separator=None,
-                        prefix_to_add="",
-                        depth=None,
-                        content_type=None):
+def __read_with_appropriate_encoding(file_path, format_):
     config_data = {}
     try:
+        with io.open(file_path, 'r', encoding='utf-8') as config_file:
+            if format_ == 'json':
+                config_data = json.load(config_file)
+
+            elif format_ == 'yaml':
+                for yaml_data in list(yaml.safe_load_all(config_file)):
+                    config_data.update(yaml_data)
+
+            elif format_ == 'properties':
+                config_data = javaproperties.load(config_file)
+                logger.debug("Importing feature flags from a properties file is not supported. If properties file contains feature flags, they will be imported as regular key-values.")
+
+    except ValueError:
```

Comment:
Yes, we also get `ValueError` if the syntax is wrong.

Reply:
I think we want to catch the very specific encoding error here.  Can we check error message or any other indicator. 

---
In reply to: [628405587](https://github.com/Azure/azure-cli/pull/17990#discussion_r628405587) [](ancestors = 628405587)

Reply:
Usually, the specific encoding error that we can catch is `UnicodeDecodeError`. It covers most cases for all file formats. But there is still a case where the presence of a BOM character in the beginning of the file doesn't throw `UnicodeDecodeError`. In JSON format, it throws `JSONDecodeError`, which is the same error that's thrown if the file syntax is incorrect. 
I am not sure if there might be any other edge cases like this. That's why I was catching the general `ValueError`.

However, I realized that we also need to catch `yaml.YAMLError` because yaml parser doesnt throw `ValueError` if the syntax is incorrect. With this change, I am catching specific errors in `__read_with_appropriate_encoding` and then general `ValueError` (if any edge cases are missed) in `__read_kv_from_file`

Reply:
What error will be thrown if BOM character exists in the yaml file? 

---
In reply to: [628514233](https://github.com/Azure/azure-cli/pull/17990#discussion_r628514233) [](ancestors = 628514233)

Reply:
None. YAML and properties files can handle the BOM character appropriately.